dali/dfu/dfuutil.cpp

```diff
@@ -857,21 +857,83 @@ class CDFUhelper: public CInterface, implements IDFUhelper
             else
                 throwError1(DFUERR_DFileNotFound, lfn);
         }
-        else {
-            Owned<IFileDescriptor> fdesc = file->getFileDescriptor();
-            Owned<IPropertyTree> t = fdesc->getFileTree();
+        else
+        {
+            Owned<IPropertyTree> t = queryDistributedFileDirectory().getFileTree(lfn, user, NULL, FOREIGN_DALI_TIMEOUT, true, true);
             toXML(t, out, true);
         }
         return out;
     }
 
-    void addFileXML(const char *lfn,const StringBuffer &xml,IUserDescriptor *user)
+    void addFileXML(const char *lfn, const StringBuffer &xml, IUserDescriptor *user)
     {
         Owned<IPropertyTree> t = createPTreeFromXMLString(xml);
-        Owned<IFileDescriptor> fdesc = deserializeFileDescriptorTree(t,&queryNamedGroupStore(),0);
-        Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fdesc,true);
-        if (file)
-            file->attach(lfn,user);
+        IDistributedFileDirectory &dfd = queryDistributedFileDirectory();
+        // Check if this XML is a superfile map
+        if (t->hasProp("@numsubfiles"))
+        {
+            Owned<ISuperFileDescriptor> fdesc = deserializeSuperFileDescriptorTree(t);
+            // It seems XML is fine to create  superfile
+            Owned<IDistributedSuperFile> superfile = dfd.createNewSuperFile(fdesc, user, false);
+
+            if (superfile)
+            {
+                unsigned numSubfiles = t->getPropInt("@numsubfiles",0);
+                if (numSubfiles)
+                {
+                    Owned<IPropertyTreeIterator> treeIter = t->getElements("SubFile");
+                    ForEach(*treeIter)
+                    {
+                        IPropertyTree & st = treeIter->get();
+                        StringBuffer subfilename;
+                        st.getProp("@name", subfilename);
+                        if (!dfd.exists(subfilename.str(), user))
+                            throw MakeStringException(-1, "Logical subfile '%s' doesn't exists!", subfilename.str());
+
+                        if (!dfd.isSuperFile(subfilename.str(), user))
+                            if (!dfd.existsPhysical(subfilename.str(), user))
+                                throw MakeStringException(-1, "Physical Parts doesn't exists for logical file %s!", subfilename.str());
+                    }
+                }
+                try
+                {
+                    superfile->attach(lfn, user);
+                    superfile->reloadSubFiles();
+                    unsigned numSubFiles = superfile->numSubFiles();
+                    for (unsigned idx=0; idx < numSubFiles; idx++)
+                        superfile->relinkSubFile(idx);
+                }
+                catch (IException *e)
+                {
+                    EXCLOG(e,"addFileXML(Super File");
+                    e->Release();
+                    dfd.removeSuperFile(lfn, true, user, 0);
```

Comment:
With the above changes, either superfile->attach() succeeds or it will fire an exception, either way you will not need to remove here.